examples/run-blockchain/index.ts

```diff
@@ -0,0 +1,117 @@
+import VM from '../../'
+
+import Account from 'ethereumjs-account'
+import * as utils from 'ethereumjs-util'
+import { promisify } from 'util'
+import PStateManager from '../../lib/state/promisified'
+import { toBuffer } from 'ethereumjs-util'
+
+const Block = require('ethereumjs-block')
+const Blockchain = require('ethereumjs-blockchain')
+const BlockHeader = require('ethereumjs-block/header.js')
+const testData = require('./test-data')
+const level = require('level')
+
+async function main() {
+  const hardfork = testData.network.toLowerCase()
+
+  const blockchain = new Blockchain({
+    hardfork,
+    // This flag can be control whether the blocks are validated. This includes:
+    //    * Verifying PoW
+    //    * Validating each blocks's difficulty, uncles, tries, header and uncles.
+    validate: true,
+  })
+
+  // When verifying PoW, setting this cache improves the performance of subsequent runs of this
+  // script. It has no effect if the blockchain is initialized with `validate: false`.
+  setEthashCache(blockchain)
+
+  const vm = new VM({
+    blockchain: blockchain,
+    hardfork,
+  })
+
+  await setupPreConditions(vm, testData)
+
+  await setGenesisBlock(blockchain, hardfork)
+
+  await putBlocks(blockchain, hardfork, testData)
+
+  await vm.runBlockchain(blockchain)
+
+  const blockchainHead = await promisify(vm.blockchain.getHead.bind(vm.blockchain))()
+
+  console.log('--- Finished processing the BlockChain ---')
+  console.log('New head:', '0x' + blockchainHead.hash().toString('hex'))
+  console.log('Expected:', testData.lastblockhash)
+}
+
+function setEthashCache(blockchain: any) {
+  if (blockchain.validate) {
+    blockchain.ethash.cacheDB = level('./.cachedb')
+  }
+}
+
+async function setupPreConditions(vm: VM, testData: any) {
+  const psm = new PStateManager(vm.stateManager)
+  for (const address of Object.keys(testData.pre)) {
+    const addressBuf = utils.toBuffer(address)
+
+    const acctData = testData.pre[address]
+    const account = new Account({
+      nonce: acctData.nonce,
+      balance: acctData.balance,
+    })
+
+    await psm.putAccount(addressBuf, account)
+
+    for (const hexStorageKey of Object.keys(acctData.storage)) {
+      const val = utils.toBuffer(acctData.storage[hexStorageKey])
+      const storageKey = utils.setLength(utils.toBuffer(hexStorageKey), 32)
+
+      await psm.putContractStorage(addressBuf, storageKey, val)
+    }
+
+    const codeBuf = utils.toBuffer(acctData.code)
+
+    await psm.putContractCode(addressBuf, codeBuf)
+  }
+
+  // This forces a cache flush, which is necessary here
+  await psm.getStateRoot()
```

Comment:
I changed this example to use the `StateManager`. It's much easier to understand now. But this line in particular (not having it), was driving me crazy. Am I missing something or is this a bug?

If removed, I get this error:

```
Error: State trie does not contain state root
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/lib/state/stateManager.ts:423:21
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/baseTrie.js:764:5
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/baseTrie.js:180:7
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/util.js:75:7
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/node_modules/async/lib/async.js:52:16
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/node_modules/async/lib/async.js:269:32
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/node_modules/async/lib/async.js:44:16
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/util.js:71:7
    at /Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/merkle-patricia-tree/baseTrie.js:157:9
    at dispatchError (/Users/pato/projects/ethereumjs/ethereumjs-vm/node_modules/levelup/lib/util.js:22:36)
```

@s1na maybe?

Reply:
I concluded that a missing cache flush was the problem when trying to print the `stateRoot` and everything worked. I may be completely confused tho.

Reply:
Hm, not sure about the cause of that error. That exception is thrown in `setStateRoot` but I don't see you using it.

Anyway if the problem is cache not being flushed, you can also use the `checkpoint` `commit` api of the StateManager.

Btw, I agree that using `StateManager` has made this much more readable, good idea!

Reply:
> Hm, not sure about the cause of that error. That exception is thrown in `setStateRoot` but I don't see you using it.

`vm.runBlockchain` calls it. The `stateRoot` is hardcoded in the test data, but missing in the actual `state`. When debugging this, I called `getStateRoot` to make check if they were different, and that's how I got into thinking that it may be a cache problem.

The `checkpoint` & `commit` thing fixed it.